sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala

```diff
@@ -397,7 +398,11 @@ trait JoinSelectionHelper extends Logging {
 
   protected def hashJoinSupported
       (leftKeys: Seq[Expression], rightKeys: Seq[Expression]): Boolean = {
-    val result = leftKeys.concat(rightKeys).forall(e => UnsafeRowUtils.isBinaryStable(e.dataType))
+    def hasStableCollationKey(e: Expression): Boolean =
+      e.dataType.isInstanceOf[StringType] || e.dataType.isInstanceOf[BinaryType]
+    def isHashJoinSupported(e: Expression): Boolean = UnsafeRowUtils.isBinaryStable(e.dataType) ||
```

Comment:
Why do you need this `hasStableCollationKey`? `isBinaryStable` will return true for binary, so things should work just work, right?